### PR TITLE
Converter for Tile operation

### PR DESCRIPTION
**Pull request by drivanov**
- ConvertTile operator is implemented so that both parameters (tensor and multiplier) can be passed as tensors and weights. A unit test with 6 tests and all combinations of passed parameters is provided.

- New macro `#define CHECK_INPUT_SIZE (size, exp_size, node_def)` unifies the testing of the number of parameters passed. It is used in 8 different places.

- The helper function template void `AdjustVectorByDims(...)` makes writing tests much easier (especially tests with parameters that can be passed as tensors and weights). For example, today we can use some parameters like tensors with predefined dimensions, but with no real numbers. An attempt to pass the same parameters with the same values as weights leads to the crash. We see a similar crash when the vector providing the values assigned to the tensor is not empty, but its length does not match the number of values defined by the dimensions. `AdjustVectorByDims` fixes these two problems and is called by default in the situations described, but these calls can also be blocked when such failures are intentional.

NOTE: This PR should be slightly adjusted if it will be merged after
[PR#54230](https://github.com/tensorflow/tensorflow/pull/54230):: Refactoring of ConvertUnary

### CODE REVIEW

**Comment by bixia1** [comment]
Remove this description from the PR description to comment:
This PR is a replacement for PR# 53698

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:1]: 2022

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1555]: Can you write a comment to describe the information printed out here?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1556]: I think it is better to move if (VLOG_IS_ON(2)) from the call sites of this routine to here.
What do you think?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1564]: Can you use absl::StrJoin and absl::StrCat to implement this functionality?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1647]: Sorry, where is this routine defined?

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:70]: If the multiple is not a weight, we don't need to perform similar check?
We need to make sure that the tensor is one dimension and its number of elements equal to the rank of inputs[0]?
is this related to line 2240-2240 in the test, where we need to fix up the error code?

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (christopherbate) [inline tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:117]: Use new DimsAdapter (defined in `tf2tensorrt/convert/utils.h`) on master top-of-tree rather than raw pointer on the dims struct.

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (christopherbate) [inline tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:123]: For manipulating `INetworkDefinition*` in below code, consider using `NetworkBuilder` in `ops/layer_utils.h`. This provides better encapsulation of network manipulation across the code base.

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (christopherbate) [inline tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:149]: You can use DimsAdapter and DimsAdapter::RemoveBatchDimension

File: tensorflow/compiler/tf2tensorrt/convert/op_converter.h
Comment (christopherbate) [inline tensorflow/compiler/tf2tensorrt/convert/op_converter.h:27]: There's another PR where these are moved to `tf2tensorrt/common/utils.h`. We should probably prefer that location since other error handeling macros are defined there.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (christopherbate) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1555]: We should prefer use of `absl::string_view` (no `const` or `&`) as function parameters over `const char*`. Same thing for `const string &` here.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (christopherbate) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:2200]: s/lenght/length

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (christopherbate) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:2189]: s/output dims/expected_output_dims/

Use correct field names for these comments. Would it hurt to label them all? For the next person to debug.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (christopherbate) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:2239]: The `k <= 2` / `k > 2` condition here seems pretty brittle. I think you are basing this on whether the `p.status.ok()` is true or false...

**Review by christopherbate — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (christopherbate) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:1829]: This check is not sufficiently general to warrant a macro. Macros should not be preferred over functions in modern C++ code base. It's hiding non-obvious behavior. In Tensorflow, `CHECK_*` macros are used for assertions. This check could be a function and called via 
`TF_RETURN_IF_ERROR(VerifyInputSizes(inputs, expected_inputs, node_def))`

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (christopherbate) [inline tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:43]: This is already being checked by the base class before this derived class' `Validate` method is called. It is determined through the defined `InputSpec` you wrote above

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1556]: I think you are right and I will make the appropriate changes

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1555]: I will do that

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1666]: Sorry, I forgot to merge it into that version. Fixed with the latest commit,

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: Done, see latest commit

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:117]: Sorry, I cannot find DimsAdapter in current version of `tf2tensorrt/convert/utils.h`

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:123]: Do you mean `TRTNetworkBuilder`?

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:149]: Sorry, I cannot find DimsAdapter in current version of `tf2tensorrt/convert/utils.h`. You seem to be mentioning brand new code that hasn't been merged yet.

File: tensorflow/compiler/tf2tensorrt/convert/op_converter.h
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/op_converter.h:27]: Sorry, I am not sure that I understand you correctly... Are you talking about adding these two lines:
```
#include "tensorflow/compiler/tf2tensorrt/convert/weights.h"
#include "third_party/tensorrt/NvInfer.h"
```
to `op_converter.h` (current version of this file does not contain them)
      OR
about moving macros to  `tf2tensorrt/common/utils.h` ?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1584]: Done. Next commit will have it

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:2215]: Done

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:2204]: Done

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:2254]: I didn't like it either. Improved.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:1829]: You are right, perhaps, the prefix CHECK_ should be replaced, but I think that the use of macros sometimes has some advantages over the C++ functions. For instance, in that case by using macro: 
```
#define CHECK_SHAPE_TENSOR(tensor)                                        \
   if (!IsTrtShapeTensorCompatible(tensor)) {                             \
     TFTRT_ERROR(errors::InvalidArgument, "Tensor of type ",              \
                 DebugString(tensor.dtype()), " having shape ",           \
                 tensor.shape().DebugString(), " is not TRT compatible"); \
   }
```
which uses an other one:
```
#define TFTRT_ERROR(func, ...)                   \
    do {                                         \
      return func("TFTRT::", __FUNCTION__, ":",  \
                    __LINE__, ": ", __VA_ARGS__);\
    } while (0)
```
we have possibility to see the exact place where this check failed, because of  ```__FUNCTION__,  __LINE__```.
We don't have this possibility with the function unless ```__FUNCTION__, __LINE__``` are used as the additional input parameters of 
```
VerifyInputSizes(..., __FUNCTION__, __LINE__)
```
By the way, `TF_RETURN_IF_ERROR` is also a macro, which use an other one: `TF_PREDICT_FALSE`

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:43]: I will remove it. It is there because my first implementation was written in the "old stile" and did not use `OpConverterBase`

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (bixia1) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:116]: You probably need to wait until the PR is merged to fix this. In the process of merging that PR.

**Comment by bixia1** [comment]
@drivanov I assume that you are still working on this as there are comments that aren't addressed yet.

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (christopherbate) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:122]: yes

File: tensorflow/compiler/tf2tensorrt/convert/op_converter.h
Comment (christopherbate) [reply tensorflow/compiler/tf2tensorrt/convert/op_converter.h:27]: moving macros. the comment is appearing above the code

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (christopherbate) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:1829]: You are right, I didn't realize we are using __LINE__ here. It still seems odd since this macro can only be used in a very particular place. I think it jus shouldn't be moved out of `convert_nodes.cc` as the converters based on the BaseOpConverter shouldn't need this class

Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:1829]: As you suggested, I renamed macros CHECK_INPUT_SIZE and  CHECK_SHAPE_TENSOR to TFTRT_CHECK_INPUT_SIZE and  TFTRT_CHECK_SHAPE_TENSOR, respectively.
Both these macros use TFTRT_ERROR.
 
As a mater of fact, now TFTRT_CHECK_SHAPE_TENSOR is used in `ops/tile.cc`. It's why now TFTRT_CHECK_SHAPE_TENSOR and TFTRT_ERROR are in `utils.h`

You are right, the other three 
- TFTRT_INTERNAL_ERROR_AT_NODE, 
- TFTRT_RETURN_ERROR_IF_NULLPTR, 
- TFTRT_CHECK_INPUT_SIZE
 
could be moved back to `convert_nodes.cc`. I will do that

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:116]: OK, I will wait this merge.

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:70]: Done, see lines 92-103 of tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:136]: Implemented the use of TRTNetworkBuilder

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc]: Implemented

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:131]: Sorry, I'm not sure how efficient it is to use `DimsAdapter` here. As far as I understand, we still need to create some intermediate vectors and cut out the first coordinates. In the current version, we do this directly.

File: tensorflow/compiler/tf2tensorrt/convert/op_converter.h
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/op_converter.h]: Done

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1585]: I understand we have made changes here, can we refine here one more time like this:

Change this routine to return a string for the output,  and name it get_debug_string.

Fix the callsites to use it like this:
if (VLOG_IS_ON(2) {
   VLOG(2) << get_debug_string_for_vector(...)
}

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:2198]: Add "." to the end of the comment.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:2192]: Please add "." to the end of the comment.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:2220]: s/number of dimentions/rank/

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:2261]: Please add "." to the end of the comment.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:2286]: This error message has a "different style" comparing with the rest, such as the one in line 2219. Can you make this one similar to the one in line 2219, by putting the values "4" and "3" right next to their description? "4" should go with "the number of elements (4)", for example.

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (bixia1) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:131]: I think you can still you DimAdapter to get "stride".
To compute dynamic_flag, can you use std::any_of?
Can you explain your expression "(size.d[i] = pMultiplies[dim_adj+i] * *(pSize+i)) < 0"?
Do you mean dynamic_flag is set to true if there is a pair in "size" and "pMultiplies", with the same negative values?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1388]: It was a merge error. Fixed.

File: tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/tile.cc:131]: 1. Done: `dynamic_flag` calculated with `std::any_of`
2. dynamic_flag is set to true when `(pMultiplies[dim_adj+i] * *(pSize+i)) < 0`. 
    Actually,  `pMultiplies[dim_adj+i] > 0` (according to corresponding check  in `ConvertTile::Validate`), therefore what we check here is just `*(pSize+i) < 0` for all i.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1388]: Sorry, it was a merge error. Fixed.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1519]: Done

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: Done

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:1585]: Done

**Comment by drivanov** [comment]
@bixia, I have addressed the issues, please have a look

**Comment by drivanov** [comment]
@bixia1:  Sorry, I have no idea how this happened. Some of these files are not even from Tensorflow but from my MXNET projects. I just fixed it.

> I see a bunch a unrelated files were uploaded, and an existing file op_convert.h is a copied to a different directory and shown as "new"? Can you please fix those?

**Comment by drivanov** [comment]
@bixia1 : I rebased before the squash, therefore I see **Conflicting files**. Is it OK?
OR, perhaps I need to create a new PR.

**Comment by bixia1** [comment]
Something is not right, it shows this PR change > 2000 files. Would you please check?

**Comment by MarkDaoust** [comment]
It might have been a wrong-way merge or something like that. I a `git rebase` is usually works for me.

At this point it's usually easier to just close it and make a new PR from scratch.

**Comment by drivanov** [comment]
@bixia1 : I just created a new [PR#4424](https://github.com/tensorflow/tensorflow/pull/54424), which should replace that one.

**Pull request closed without merging**
